### PR TITLE
CSP layout selector

### PR DESCRIPTION
**Pull request by 1ucian0**
This PR models the layout selection problem as a constraint satisfaction problem.

### CODE REVIEW

**Comment by 1ucian0** [comment]
With our benchmarks? Sure.. let me try.

**Comment by 1ucian0** [comment]
I tried with
```python
cmap20 = CouplingMap(FakeTokyo().configuration().coupling_map)
nqubits = [4, 8, 12, 16, 20]
depths = [16, 32, 64, 128, 256, 512, 1024, 2048, 4096]

for nqubit in nqubits:
    for depth in depths:
        circuit = random_circuit(nqubit, depth)
        dag = circuit_to_dag(circuit)
        pass_ = CSPLayout(cmap20)
        import time

        start = time.process_time()
        pass_.run(dag)
        print('nqubit:', nqubit, 'depth:', depth, time.process_time() - start)
        print('None' if pass_.property_set['layout'] is None else 'not None')
```
After `nqubit: 4 depth: 4096` (it took 0.04 seconds), it cannot find any solution. The worst situation was half second. I think it can be a good attempt before applying `TrivialLayout`.

**Comment by 1ucian0** [comment]
`python-constraint` is a simple pure python lib. Z3 is a full different animal. I can explore Z3 for more hard core options. But still think that this approach is very light-weight and it should be considered before `TrivialLayout`. At least for some optimization levels.

**Comment by yehuda-naveh** [comment]
If at some stage we want to go heavyweight with this, we have a couple of in-house of CSP and SAT engines at our disposal. One comment though - both CSP and SAT are supposed exponential in the number of variables, but linear in number of constraints - in fact adding constraints many times greatly reduce runtime to solution because it efficiently prunes the search tree. Also, in our case the challenge is optimization - not satisfaction per se, so may need to consider engines which deal efficiently with soft constraints hierarchies. The other option is to do a log-search with adding an upper-bound constraint after each run, but this is a constraint CSP engines don't deal with very well. Z3 has an interface to soft constraints, but I never checked to what extent it can actually deal with them

**Comment by 1ucian0** [comment]
> I haven't reviewed all the code in depth yet, but the release notes (which are great details thanks for writing them) are in the wrong format.

Ups! Fixed in https://github.com/Qiskit/qiskit-terra/pull/3043/commits/b445b41178e7ca78fc10b1566f47a666da86c0de

File: qiskit/transpiler/passes/mapping/csp_layout.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A pass for choosing a Layout of a circuit onto a Coupling graph, as a
+Constraint Satisfaction Problem.
```
Comment (ajavadia) [inline qiskit/transpiler/passes/mapping/csp_layout.py]: just add a sentence here to say that it either finds a layout that satisfies all swaps in a circuit, or it gives up and does not select a layout at all.

File: requirements.txt [hunk 2]
```diff
@@ -7,3 +7,4 @@ ply>=3.10
 psutil>=5
 scipy>=1.0
 sympy>=1.3
+python-constraint
```
Comment (ajavadia) [inline requirements.txt]: this pass looks good to me, but it's adding an extra dependency for something that is not even invoked during the standard compilation/execution flow. can this be made optional or upon-request or something?

File: qiskit/transpiler/passes/mapping/csp_layout.py (hunk 1 above)
Comment (1ucian0) [reply qiskit/transpiler/passes/mapping/csp_layout.py]: Done in 23a20b6e

File: requirements.txt (hunk 2 above)
Comment (1ucian0) [reply requirements.txt]: I think it makes sense to put in the default pipeline, since its relatively fast. However, we can discuss it later.

Comment (1ucian0) [reply requirements.txt]: done in 5faf45c7

**Pull request merged**